deepsubtree_test.go

```diff
@@ -232,18 +240,165 @@ func TestDeepSubtreeWWithAddsAndDeletes(t *testing.T) {
 
 	// Delete all the keys we added and check root hashes stay equal
 	for i := range keysToAdd {
-		keyToAdd := keysToAdd[i]
-		dst.Remove(keyToAdd)
-		dst.SaveVersion()
-		rootHash, err := dst.WorkingHash()
+		keyToDelete := keysToAdd[i]
+
+		existenceProofs, err := tree.getExistenceProofsNeededForRemove(keyToDelete)
+		require.NoError(err)
+		rootHash, err := tree.WorkingHash()
 		require.NoError(err)
-		err = dst.BuildTree(rootHash)
+		err = dst.AddExistenceProofs(existenceProofs, rootHash)
 		require.NoError(err)
-		tree.Remove(keyToAdd)
+
+		dst.Remove(keyToDelete)
+		dst.SaveVersion()
+		tree.Remove(keyToDelete)
 		tree.SaveVersion()
 
 		areEqual, err := haveEqualRoots(dst.MutableTree, tree)
 		require.NoError(err)
 		require.True(areEqual)
 	}
 }
+
+func readByte(r *bytes.Reader) byte {
+	b, err := r.ReadByte()
+	if err != nil {
+		return 0
+	}
+	return b
+}
+
+func FuzzBatchAddReverse(f *testing.F) {
```

Comment:
I know it's just a test, but this function is huge and complex. Can we for example:
* extract `key` as a function, instead of using lambda? 
* have separate helper functions for `Set` and `Remove`?
* probably introducing a struct with methods will make most sense (so we don't need to pass to many params)

Reply:
Could you please clarify "probably introducing a struct with methods will make most sense (so we don't need to pass to many params)" in the updated code?

Reply:
Something like:
```go
type testContext struct {
  r io.Reader
  tree *MutableTree
  dst *DeepSubTree
  keys set.Set[string]
}

func (tc *testContext) getKey(getRandom bool) ([]byte, error) {
// ...
}

func (tc *testContext) removeInDst(key []byte) error {
// ...
}

func (tc *testContext) setInDst([key, value []byte, isNewKey, isFirstKey bool) error {
// .... 
}
```

Reply:
Thanks, updated!